nanocoat/lib/scritchui/scritchPencilRasterShapes.c

```diff
@@ -17,7 +17,399 @@
+
+	/* Failed? */
+	if (sjme_error_is(error))
+		goto fail_any;
+	
+	/* Success? */
+	return error;
+	
+fail_any:
+	
+	return sjme_error_default(error);
+}
+
+sjme_errorCode sjme_scritchpen_corePrim_fillPolygon(
+	sjme_attrInNotNull sjme_scritchui_pencil g,
+	sjme_attrInValue sjme_jint xPoints[],
+	sjme_attrInPositive sjme_jint xOffset,
+	sjme_attrInValue sjme_jint yPoints[],
+	sjme_attrInPositive sjme_jint yOffset,
+	sjme_attrInPositive sjme_jint nPoints)
+{
+	sjme_errorCode error;
+	sjme_scritchui_pencilDrawHorizFunc drawHoriz;
+	int ymin = 2147483647, ymax = -2147483648, intersectionCount = 0, xStart,
+	xEnd, temp, dy, j;
+	int intersections[nPoints];
+	long ix;
+	sjme_scritchui_line* clipLine;
+
+	if (g == NULL)
+		return SJME_ERROR_NULL_ARGUMENTS;
+
+	error = SJME_ERROR_NONE;
+
+	if (!g->state.applyAlpha && g->impl->drawHorizSrc != NULL)
+		drawHoriz = g->impl->drawHorizSrc;
+	else if (g->state.applyAlpha && g->impl->drawHorizSrcOver != NULL)
+		drawHoriz = g->impl->drawHorizSrcOver;
+	else
+		drawHoriz = g->prim.drawHoriz;
+
+	/* Get clipping information. */
+	clipLine = &g->state.clipLine;
+
+	/**
+	 * Filling polygons is done through the canonical Scan Line fill algorithm. 
+	 * It works just like its description: Find the yMax and yMin of the
+	 * polygon, calculate the intersections between each edge, sort 
+	 * intersections by increasing X coordinate, then fill from top to bottom.
+	 */
+
+	for (int i = 0; i < nPoints; i++) 
+	{
+		if (yPoints[i+yOffset] < ymin) { ymin = yPoints[i+yOffset]; }
+		if (yPoints[i+yOffset] > ymax) { ymax = yPoints[i+yOffset]; }
+	}
+
+	/* Clip ymin and ymax to the screen area if any vertex is outside */
+	if(ymin + g->state.translateReal.y < clipLine->s.y) 
+		ymin = clipLine->s.y - g->state.translateReal.y;
+	
+	if(ymax + g->state.translateReal.y >= clipLine->e.y) 
+		ymax = clipLine->e.y - g->state.translateReal.y;
+
+	for (int y = ymin; y < ymax; y++)
+	{
+		intersectionCount = 0;
+		for (int i = 0; i < nPoints; i++)
+		{
+			j = (i + 1) % nPoints;
+			if ((yPoints[i + yOffset] <= y && yPoints[j + yOffset] > y) ||
+				(yPoints[j + yOffset] <= y && yPoints[i + yOffset] > y))
+			{
+				dy = yPoints[j + yOffset] - yPoints[i + yOffset];
+				if (dy != 0)
+				{
+					ix = (long) xPoints[i + xOffset] * dy + (y - yPoints[i +
+							yOffset]) * (long) (xPoints[j + xOffset] -
+							xPoints[i + xOffset]);
+					ix /= dy;
+					intersections[intersectionCount++] = (int) ix;
+				}
+			}
+		}
+
+		for (int i = 0; i < intersectionCount - 1; i++)
+		{
+			for (int j = 0; j < intersectionCount - 1 - i; j++)
+			{
+				if (intersections[j] > intersections[j + 1])
+				{
+					temp = intersections[j];
+					intersections[j] = intersections[j + 1];
+					intersections[j + 1] = temp;
+				}
+			}
+		}
+
+		for (int i = 0; i < intersectionCount; i += 2)
+		{
+			if (i + 1 < intersectionCount)
+			{
+				xStart = fmax(intersections[i], clipLine->s.x);
+				xEnd = fmin(intersections[i + 1], clipLine->e.x);
```

Comment:
Needless conversion to/from `float`, this would be better using `min/max`, if missing those can be added in `util.h`.